mono/metadata/marshal.c

```diff
@@ -6223,19 +6223,29 @@ MonoObjectHandle
 mono_icall_handle_new (gpointer rawobj)
 {
 #ifdef MONO_HANDLE_TRACK_OWNER
-	return mono_handle_new (rawobj, "<marshal args>");
+	return MONO_HANDLE_NEW (MonoObject, (MonoObject*)rawobj, "<marshal args>");
 #else
-	return mono_handle_new (rawobj);
+	return MONO_HANDLE_NEW (MonoObject, (MonoObject*)rawobj);
 #endif
 }
 
 MonoObjectHandle
 mono_icall_handle_new_interior (gpointer rawobj)
 {
 #ifdef MONO_HANDLE_TRACK_OWNER
-	return mono_handle_new_interior (rawobj, "<marshal args>");
+	MonoRawHandle const rawh = mono_handle_new_interior (rawobj, "<marshal args>");
+#else
+	MonoRawHandle const rawh = mono_handle_new_interior (rawobj);
+#endif
+#if MONO_TYPE_SAFE_HANDLES
+	// This is the only use of mono_handle_new_interior.
+	// It lacks the macro support of mono_handle_new.
+	//return *(MonoObjectHandle*)&rawh; // violates gcc -fstrict-aliasing
+	// Legal with gcc -fstrict-aliasing:
+	union { MonoRawHandle rawh; MonoObjectHandle objh; } type_pun = { rawh };
+	return type_pun.objh;
 #else
-	return mono_handle_new_interior (rawobj);
+	return rawh;
```

Comment:
You know, I think `mono_icall_handle_new_interior` has the wrong return type.  It should probably be `MonoRawHandle` - `mono_handle_new_interior` doesn't really return a handle to a MonoObject.  It returns a handle to a pointer into the middle of some valuetype.

```
MonoRawHandle mono_icall_handle_new_interior (gpointer rawptr);
```

Reply:
You mean change MonoRawHandle to gpointer? Same thing but ok. Separate change?

Reply:
No I mean declare `MonoRawHandle mono_icall_handle_new_interior (gpointer rawobj);` and then implement it as

```c
MonoRawHandle
mono_icall_handle_new_interior (gpointer rawobj)
{
#ifdef MONO_HANDLE_TRACK_OWNER
    return mono_handle_new_interior (rawobj, "<marshal args>");
#else
    return mono_handle_new_interior (rawobj);
#endif
}
```

The only reason you have to jump through hoops here is because the icall is declared with a poor (incorrect) choice of return type.  So just change the return type.

Reply:
Got it. MonoRawHandle should probably be gpointer or gpointer* then.
gpointer* is more correct, gpointer is more opaque.